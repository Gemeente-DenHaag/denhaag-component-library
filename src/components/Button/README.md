# Button

![npm (scoped)](https://img.shields.io/npm/v/@gemeente-denhaag/button?logo=npm&style=flat-square)
![GitHub Workflow Status (branch)](https://img.shields.io/github/workflow/status/Gemeente-Denhaag/denhaag-component-library/Build%20and%20deploy%20Storybook%20to%20Azure%20Web%20App/master?logo=github&style=flat-square)

Buttons are used for actions such as "Add", "Close", "Cancel" or "Save".
Button labels indicate what action will take place when the user interacts with it.

## When to use

Use buttons to communicate actions users can take and to allow users to interact with the page.
Each page should have one primary button, and any remaining calls to action should be represented as lower emphasis buttons.

## Alternatives and related components

**Use links** for navigation actions that appear within or immediately after a sentence.

## Anatomy

The button consists of:

1. Label: communicates the action that will be performed when the user interacts with it.
2. Icon (optional): provides support for the label.
3. Container

## (Interactive) states

The button contains the states normal, hover, disabled and focus.

## Design properties

### Typography

_Large_

- Label: TheSans/lg/400

_Medium_

- Label: TheSans/md/400

### Colors

_Primary button_

- Label: text color white
- Icon: svg color white
- Container: background-color Green/3

_Secondary button_

- Label: text color Green/3
- Icon: svg color Green/3
- Container: background-color white, border Green/3

### Interactive states

_Primary button_

- Hover: container background-color Green/4
- Disabled: container background-color Grey/2
- Focus: container border color Ocher/5

_Secondary button_

- Hover: label text color Green/4, icon svg color Green/4, container border color Green/4
- Disabled: label text color Grey/2, icon svg color Grey/2, container border color Grey/2
- Focus: container border color Ocher/5

### Structure

_Large_

- Label: padding-left and padding-right 20px
- Icon: height and width 24px
- Icon-left: padding-left 16px, padding-right 8px
- Icon-right: padding-left 8px, padding-right 16px
- Container: height 48px

_Medium_

- Label: padding-left and padding-right 16px
- Icon: height and width 20px
- Icon-left: padding-left 12px, padding-right 8px
- Icon-right: padding-left 8px, padding-right 12px
- Container: height 40px

_Primary button_

- Container focus: container border 2px, dash 5,5

_Secondary button_

- Container: border 1px
- Container focus: container border 2px, dash 5,5

## Accessibility

Alerts provide context and assist workflows for users with disabilities.

- Alerts use a combination of icons and colors to show their meaning and importance.
- [technical accessibility requirements]

### Navigation

Users generally expect buttons to send data or take action, and to be able to navigate links. If navigation is required for the button component, use the url prop. The control displays an anchor in the form of a button, instead of a button in HTML, to convey this difference.

See links for more information on creating accessible links.

### Labeling

[technical requirements, such as: accessibilityLabel]

### External links

[technical requirements]

### Keyboard support

Buttons use the browser's default settings for keyboard interactions.

- Give buttons keyboard focus with the tab key (or shift + tab when tabbing backwards)
- Activate buttons with the enter / return key or the space key

## Content guidelines

Buttons should be clear and predictable. Users should be able to anticipate what will happen when they select a button. Never mislead anyone by mislabeling a button.

Buttons should always lead with a strong verb that encourages action. To provide users with enough context, you can use the content formula {verb} + {noun} on buttons, except in the case of general actions such as "Done", "Close", "Cancel" or "OK".

## Best practices

### Do's

Buttons must:

- Be clearly and accurately labeled.
- Contain a strong and action-oriented verb.
- Using the right button colors in the right context. For example, only use a red error button for an action that is difficult or impossible to undo.
- Give priority to the most important actions. Too many calls to action can cause confusion and make users unsure of what to do next.
- Be positioned in consistent locations in the interface.

### Don'ts

Buttons must:

- Do not guess the user what a button does, especially if it is irreversible.
- Not be used for navigation. You can use links for this.
- Not wrapped on more than 1 line. Make sure the text stays on one line.

### Buttons versus links

Buttons are mainly used for actions such as "Add", "Close", "Cancel" or "Save". Regular buttons similar to links are used for less important or less frequently used actions, such as "View details".

Links are mainly used for navigation and usually appear in or immediately after a sentence.

The HTML displayed for the Button and Link components has meaning. Using these components intentionally and consistently results in:

- a more inclusive experience for assistive technology users
- a more coherent visual experience for sighted users
- products that are easier to maintain to scale

## References

https://uxmovement.com/buttons/5-rules-for-choosing-the-right-words-on-button-labels/ <br/>
https://uxplanet.org/7-basic-rules-for-button-design-63dcdf5676b4 <br/>
https://www.invisionapp.com/inside-design/comprehensive-guide-designing-ux-buttons/
