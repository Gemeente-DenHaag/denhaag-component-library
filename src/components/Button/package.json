{
  "name": "@gemeente-denhaag/button",
  "description": "A Button component",
  "author": "Municipality of The Hague",
  "license": "EUPL-1.2",
  "version": "0.1.7",
  "main": "dist/index.js",
  "module": "dist/index.js",
  "exports": "./dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc -b && copyfiles -u 1 \"src/**/*.css\" dist",
    "clean": "yarn rimraf dist tsconfig.tsbuildinfo"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/Gemeente-DenHaag/denhaag-component-library.git",
    "directory": "src/components/Button"
  },
  "bugs": "https://github.com/Gemeente-DenHaag/denhaag-component-library/issues",
  "dependencies": {
    "@gemeente-denhaag/baseprops": "^0.1.6",
    "@gemeente-denhaag/design-tokens-common": "^0.1.0",
    "@gemeente-denhaag/design-tokens-components": "^0.1.0",
    "@material-ui/core": "^4.11.0"
  },
  "devDependencies": {
    "@material-ui/icons": "^4.11.2"
  },
  "peerDependencies": {
    "react": "^17.0.1"
  },
  "gitHead": "dcf72a9b79266c1ebede35aff4a02dd9121a980f"
}
