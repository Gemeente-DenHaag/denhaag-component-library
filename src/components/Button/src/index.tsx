import MaterialButton from '@material-ui/core/Button';
import React from 'react';
import BaseProps from '@gemeente-denhaag/baseprops';

export interface ButtonProps extends BaseProps {
  /**
   * Simple click handler
   */
  onClick?: () => void;

  /**
   * Color for the component
   */
  color?: 'primary' | 'secondary' | 'default';

  /**
   * The url to link to when the button is clicked.
   * If defined, an a element will be used as the root node.
   */
  href?: string;

  /**
   * Size of the component
   */
  size?: 'small' | 'medium' | 'large';

  /**
   * Disables Button
   */
  disabled?: boolean;

  /**
   * Button variant
   */
  variant?: 'outlined' | 'contained' | 'text';

  /**
   * HTML type prop
   */
  type?: 'button' | 'submit' | 'reset';
}

/**
 * Primary UI component for user interaction
 */
export const Button: React.FC<ButtonProps> = (props: ButtonProps) => {
  return <MaterialButton {...props}>{props.children}</MaterialButton>;
};

/**
 * Default export for Button
 */
export default Button;
