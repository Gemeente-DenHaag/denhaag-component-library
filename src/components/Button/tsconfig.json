{
  "extends": "../../../tsconfig.build.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "dist",
    "tsBuildInfoFile": "./dist/tsconfig.tsbuildinfo",
  },
  "include": ["src/**/*"],
  "exclude": ["dist", "src/stories/*.stories.tsx"],
  "references": [
    {
      "path": "../BaseProps"
    }
  ]
}
