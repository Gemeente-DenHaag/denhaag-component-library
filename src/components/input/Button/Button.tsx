import MaterialButton from '@material-ui/core/Button'
import React from 'react'
import BaseProps from '../../BaseProps/BaseProps'
import './button-design-tokens.module.css'
// import { createMuiTheme, Theme, ThemeProvider } from "@material-ui/core";

export interface ButtonProps extends BaseProps {
  /**
   * Simple click handler
   */
  onClick?: () => void

  /**
   * Color for the component
   */
  color?: 'primary' | 'secondary' | 'default'

  /**
   * The url to link to when the button is clicked.
   * If defined, an a element will be used as the root node.
   */
  href?: string

  /**
   * Size of the component
   */
  size?: 'small' | 'medium' | 'large'

  /**
   * Disables Button
   */
  disabled?: boolean

  /**
   * Button variant
   */
  variant?: 'outlined' | 'contained' | 'text'
}

/**
 * Primary UI component for user interaction
 */
export const Button: React.FC<ButtonProps> = (
  props: ButtonProps
) => {
  return (
      <MaterialButton {...props}>{props.children}</MaterialButton>
  )
}

/**
 * Default export for Button
 */
export default Button
