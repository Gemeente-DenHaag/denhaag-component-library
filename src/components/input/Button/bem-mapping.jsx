export const bem = {
  root: 'mdh-button',
  label: 'mdh-button__label',
  contained: 'mdh-button--primary-action',
  outlined: 'mdh-button--secondary-action',
  disabled: 'mdh-button--disabled',
}
