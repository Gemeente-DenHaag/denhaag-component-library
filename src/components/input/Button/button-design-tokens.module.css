:root {
  --mdh-button-background-color: rgb(34, 123, 60);
  --mdh-button-secondary-background-color: #EDBF07;
  --mdh-button-focus-background-color: #9a6f1e;
  --mdh-button-focus-text-color: white;
  --mdh-button-hover-background-color: #9a6f1e;
  --mdh-button-hover-text-color: white;
  --mdh-button-disabled-background-color: #666;
  --mdh-button-font-family: "TheMix", sans-serif;
}

.mdh-button {
  background-color: var(--mdh-button-background-color);
  font-family: var(--mdh-button-font-family);
}

.mdh-button--focus, .mdh-button:focus {
  background-color: var(--mdh-button-focus-background-color);
  color: var(--mdh-button-focus-text-color);
}

.mdh-button--hover, .mdh-button:hover {
  background-color: var(--mdh-button-hover-background-color);
  color: var(--mdh-button-hover-text-color);
}

.mdh-button--disabled {
  background-color: var(--mdh-button-disabled-background-color);
}
