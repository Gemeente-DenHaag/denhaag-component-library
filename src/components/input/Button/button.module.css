.mdh-button {
  --mdh-button-padding-vertical: 0.6875rem;
  --mdh-button-padding-horizontal: 1.25rem;
  --mdh-button-border-radius: 0.1875rem;
  --mdh-button-focus-border-color: #CCA000;
  --mdh-button-focus-border-width: 0.125rem;
  --mdh-button-font-family: "TheMix", sans-serif;

  --mdh-button-primary-action-background-color: #248641;
  --mdh-button-primary-action-color: #FFFFFF;
  --mdh-button-primary-action-hover-background-color: #1D6B34;
  --mdh-button-primary-action-hover-color: #FFFFFF;
  --mdh-button-primary-action-disabled-background-color: #D2D2D2;
  --mdh-button-primary-action-disabled-color: #FFFFFF;

  --mdh-button-secondary-action-background-color: #FFFFFF;
  --mdh-button-secondary-action-color: #248641;
  --mdh-button-secondary-action-border-color: #248641;
  --mdh-button-secondary-action-border-width: 0.0625rem;
  --mdh-button-secondary-action-hover-background-color: #FFFFFF;
  --mdh-button-secondary-action-hover-color: #1D6B34;
  --mdh-button-secondary-action-hover-border-color: #1D6B34;
  --mdh-button-secondary-action-disabled-background-color: #FFFFFF;
  --mdh-button-secondary-action-disabled-color: #D2D2D2;
  --mdh-button-secondary-action-disabled-border-color: #D2D2D2;

  padding: var(--mdh-button-padding-vertical) var(--mdh-button-padding-horizontal);
  border-radius: var(--mdh-button-border-radius);
  box-shadow: none;
  transition: none;
}

.mdh-button--primary-action {
  background-color: var(--mdh-button-primary-action-background-color);
  color: var(--mdh-button-primary-action-color);
}

.mdh-button--primary-action.mdh-button--hover, .mdh-button--primary-action:hover {
  background-color: var(--mdh-button-primary-action-hover-background-color);
  color: var(--mdh-button-primary-action-hover-color);
}

.mdh-button--primary-action.mdh-button--disabled {
  background-color: var(--mdh-button-primary-action-disabled-background-color);
  color: var(--mdh-button-primary-action-disabled-color);
}

.mdh-button--secondary-action {
  background-color: var(--mdh-button-secondary-action-background-color);
  color: var(--mdh-button-secondary-action-color);
  border-width: var(--mdh-button-secondary-action-border-width);
  border-style: solid;
  border-color: var(--mdh-button-secondary-action-border-color);
}

.mdh-button--secondary-action.mdh-button--hover, .mdh-button--secondary-action:hover {
  background-color: var(--mdh-button-primary-action-hover-background-color);
  color: var(--mdh-button-secondary-action-hover-color);
  border-color: var(--mdh-button-secondary-action-hover-border-color);
}

.mdh-button--secondary-action.mdh-button--disabled {
  background-color: var(--mdh-button-primary-action-disabled-background-color);
  color: var(--mdh-button-secondary-action-disabled-color);
  border-color: var(--mdh-button-secondary-action-disabled-border-color);
}

.mdh-button--hover, .mdh-button:hover {
  box-shadow: none;
}

.mdh-button--focus, .mdh-button:focus {
  border-width: var(--mdh-button-focus-border-width);
  border-style: solid;
  border-color: var(--mdh-button-focus-border-color);
  box-shadow: none;
}


/* EXAMPLE FOR HOW TO STYLE DEEPER INTO THE REACT COMPONENTS */
.mdh-button__label {
  font-size: 2rem;
}
