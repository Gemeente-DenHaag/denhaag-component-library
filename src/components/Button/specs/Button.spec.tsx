import * as React from 'react';
import { mount } from '@cypress/react';
import Button, { ButtonProps } from '../src/index';

describe('Button tests', () => {
  it('can be clicked', () => {
    const props: ButtonProps = {
      onClick: () => {
        console.log('Clicked');
      },
    };

    const spy = cy.spy(props, 'onClick').as('Button click');

    mount(<Button onClick={props.onClick}>Test button</Button>);

    cy.get('button')
      .contains('Test button')
      .click()
      .then(() => expect(spy).to.have.been.calledOnce);
  });

  it('does not violate any accessibility rules', () => {
    mount(<Button>Test button</Button>);
    cy.injectAxe();
    cy.checkA11y('button');
  });

  it('matches snapshot', () => {
    mount(<Button>Test button</Button>);
    cy.get('button').toMatchSnapshot();
  });
});
