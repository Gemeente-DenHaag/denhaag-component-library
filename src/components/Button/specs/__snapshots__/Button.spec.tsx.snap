exports[`Button tests > matches snapshot #0`] = `
<button
  class="MuiButtonBase-root MuiButton-root denhaag-button MuiButton-contained denhaag-button--primary-action denhaag-button--medium"
  tabindex="0"
  type="button"
>
  <span class="MuiButton-label denhaag-button__label">Test button</span>
</button>
`;
